### PR TITLE
created shopping cart and cart item entities and added implementation…

### PR DESCRIPTION
**Pull request by nastasiiaKulinich**
… for necessary functionality

### CODE REVIEW

File: src/main/java/com/example/bookstore/config/OpenApiConfig.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class OpenApiConfig {
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+                .components(new Components().addSecuritySchemes("BearerAuth",
+                        new SecurityScheme()
+                                .type(SecurityScheme.Type.HTTP)
+                                .scheme("bearer")
+                                .bearerFormat("JWT")))
+                .addSecurityItem(new SecurityRequirement().addList("BearerAuth"));
```
Comment (MentorMateAcademy) [inline src/main/java/com/example/bookstore/config/OpenApiConfig.java]: Strings can save to constants

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.addBookToShoppingCart(user.getId(), cartItemRequestDto);
+    }
+
+    @ResponseStatus(HttpStatus.OK)
+    @PutMapping("/items/{cartItemId}")
+    @Operation(summary = "Update a cart item",
+            description = "Update quantity of a book in the shopping cart")
+    @PreAuthorize("hasRole('USER')")
+    public CartItemDto updateCartItem(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItem(
```

File: src/main/java/com/example/bookstore/dto/cartitem/CreateCartItemRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CreateCartItemRequestDto {
+    @NotNull
+    @Positive
+    private Long bookId;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/dto/cartitem/CreateCartItemRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/dto/cartitem/UpdateCartItemRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.example.bookstore.dto.cartitem;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class UpdateCartItemRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/dto/cartitem/UpdateCartItemRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/mapper/ShoppingCartMapper.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import com.example.bookstore.model.ShoppingCart;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.Mappings;
+
+@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
+public interface ShoppingCartMapper {
+    @Mappings({
+            @Mapping(source = "user.id", target = "userId"),
+            @Mapping(source = "cartItems", target = "cartItems")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/mapper/ShoppingCartMapper.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/mapper/ShoppingCartMapper.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    })
+    ShoppingCartDto toDto(ShoppingCart shoppingCart);
+
+    @Mappings({
+            @Mapping(source = "userId", target = "user.id"),
+            @Mapping(source = "cartItems", target = "cartItems")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/mapper/ShoppingCartMapper.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+@Setter
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Entity
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/bookstore/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/bookstore/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    private Long id;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        ShoppingCart shoppingCart = getShoppingCartByUserId(id);
+        CartItem cartItem = cartItemMapper.toEntity(cartItemRequestDto);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public CartItemDto updateCartItem(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartDto updateCartItem(
```

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    public ShoppingCartDto getShoppingCart(Long id) {
+        ShoppingCart shoppingCart = getShoppingCartByUserId(id);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addBookToShoppingCart(
+            Long id, CreateCartItemRequestDto cartItemRequestDto) {
+        ShoppingCart shoppingCart = getShoppingCartByUserId(id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:40]: Find book in DB first
Check if book already exists in shop cart. If so -> increase quantity, if no -> add cart item

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        cartItem.setQuantity(cartItemRequestDto.getQuantity());
+        CartItem savedCartItem = cartItemRepository.save(cartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    @Override
+    @Transactional
+    public void removeCartItem(Long cartItemId) {
+        CartItem cartItem = getCartItemById(cartItemId);
+        cartItemRepository.delete(cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:73]: Be sure, that cartItem will be removed not only from DB but also from shop cart Set<CartItem> cartItems

File: src/main/java/com/example/bookstore/service/user/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -32,9 +34,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         User user = userMapper.toModel(requestDto);
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
         Set<Role> roles = new HashSet<>();
-        roles.add(roleRepository.findRoleByName(Role.RoleName.USER));
+        roles.add(roleRepository.findRoleByName(Role.RoleName.ADMIN));
         user.setRoles(roles);
         User savedUser = userRepository.save(user);
+        shoppingCartService.registerNewShoppingCart(savedUser);
         return userMapper.toDto(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/user/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
        shoppingCartService.registerNewShoppingCart(user);
        return userMapper.toDto(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job. Let's improve it

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**